src/charm.py

```diff
@@ -90,150 +139,158 @@ def __init__(self, *args):
-            new_content=gnb_config_content,
-        ):
-            self._push_gnb_config_file_to_workload(gnb_config_content)
+        plan = self._container.get_plan()
+        if plan.services != self._pebble_layer.services:
+            self._container.add_layer(self._container_name, self._pebble_layer, combine=True)
+            self._container.replan()
+            logger.info("New layer added: %s", self._pebble_layer)
+        if restart:
+            self._container.restart(self._service_name)
+            logger.info("Restarted container %s", self._service_name)
+            return
 
-    def _get_existing_config_file(self, path: str) -> str:
-        """Get the existing config file from the workload.
+    def _is_config_update_required(self, content: str) -> bool:
+        return not self._webui_config_file_exists() or not self._webui_config_file_content_matches(
+            content=content)
 
-        Args:
-            path (str): Path to the config file.
+    def _webui_config_file_content_matches(self, content: str) -> bool:
+        if not self._webui_config_file_exists():
+            return False
+        existing_content = self._container.pull(path=WEBUI_CONFIG_PATH)
+        return existing_content.read() == content
+
+    def _webui_config_file_exists(self) -> bool:
+        return bool(self._container.exists(WEBUI_CONFIG_PATH))
+
+    def _generate_webui_config_file(self) -> str:
+        return render_config_file(
+            common_database_name=COMMON_DATABASE_NAME,
+            common_database_url=self._get_common_database_url(),
+            auth_database_name=AUTH_DATABASE_NAME,
+            auth_database_url=self._get_auth_database_url(),
+        )
 
-        Returns:
-            str: Content of the config file.
-        """
-        if self._container.exists(path=path):
-            existing_content_stringio = self._container.pull(path=path)
-            return existing_content_stringio.read()
-        return ""
+    def _is_nms_service_running(self) -> bool:
+        if not self._container.can_connect():
+            return False
+        try:
+            service = self._container.get_service(self._service_name)
+        except ModelError:
+            return False
+        return service.is_running()
 
-    def _get_upf_host_port_list(self) -> List[Tuple[str, int]]:
-        """Get the list of UPF hosts and ports from the `fiveg_n4` relation data bag.
+    def _configure_upfs(self) -> None:
+        if not self.model.relations.get(FIVEG_N4_RELATION_NAME):
+            logger.info("Relation %s not available", FIVEG_N4_RELATION_NAME)
+        webui_upf_config = self._webui.get_upfs()
+        relation_upf_config = self._get_upf_config_from_relations()
+        self._sync_upfs(webui_upfs=webui_upf_config, relation_upfs=relation_upf_config)
 
-        Returns:
-            List[Tuple[str, int]]: List of UPF hostnames and ports.
-        """
+    def _configure_gnbs(self) -> None:
```

Comment:
I don't really understand the current function separation between `configure_gnbs` and `sync_gnbs`. I think we can simplify the code a bit by only having a `sync_gnbs()` (taking 0 arguments), calling it directly from `_configure_sdcore_nms` and have the whole logic in `sync_gnbs`.